crates/ruff/src/rules/pylint/rules/repeated_equality_comparison_target.rs

```diff
@@ -0,0 +1,150 @@
+use crate::checkers::ast::Checker;
+use itertools::any;
+use ruff_diagnostics::{Diagnostic, Violation};
+use ruff_macros::{derive_message_formats, violation};
+use ruff_python_ast::hashable::HashableExpr;
+use rustc_hash::FxHashMap;
+use rustpython_parser::ast::{BoolOp, CmpOp, Expr, ExprCompare, Ranged};
+use std::ops::Deref;
+
+/// ## What it does
+/// Checks for equality comparisons that can rewritten as a membership test.
+///
+/// ## Why is this bad?
+/// Instead, to check if a variable is equal to one of many values, combine the
+/// values into a collection and use the `in` operator. This is faster and less
+/// verbose.
+///
+/// If the items are hashable, use a `set` instead of a `list`. Membership
+/// using the `in` operator is more efficient for sets than for non-hashable
+/// collections like lists and tuples.
+///
+/// ## Example
+/// ```python
+/// foo == "bar" or foo == "baz" or foo == "qux"
+/// ```
+///
+/// Use instead:
+/// ```python
+/// foo in {"bar", "baz", "qux"}
+/// ```
+///
+/// ## References
+/// - [Python documentation: Comparisons](https://docs.python.org/3/reference/expressions.html#comparisons)
+/// - [Python documentation: Membership test operations](https://docs.python.org/3/reference/expressions.html#membership-test-operations)
+/// - [Python documentation: `set`](https://docs.python.org/3/library/stdtypes.html#set)
+#[violation]
+pub struct RepeatedEqualityComparisonTarget {
+    expr: String,
+}
+
+impl Violation for RepeatedEqualityComparisonTarget {
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        let RepeatedEqualityComparisonTarget { expr } = self;
+        format!(
+            "Consider merging multiple comparisons: `{expr}`. \
+            Use a `set` if the elements are hashable."
+        )
+    }
+}
+
+/// PLR1714
+pub(crate) fn repeated_equality_comparison_target(
+    checker: &mut Checker,
+    expr: &Expr,
+    op: BoolOp,
+    values: &[Expr],
+) {
+    if !op.is_or() && !op.is_and() {
+        return;
+    }
+    if any(values.iter(), |v| !is_allowed_value(op, v)) {
+        return;
+    }
+    let mut left_to_comparators: FxHashMap<HashableExpr, (usize, Vec<HashableExpr>)> =
+        FxHashMap::default();
+    for value in values {
+        match value {
+            Expr::Compare(ExprCompare {
+                left, comparators, ..
+            }) => {
+                let (count, matches) = left_to_comparators
+                    .entry(left.deref().into())
+                    .or_insert_with(|| (0, Vec::new()));
+                *count += 1;
+                matches.extend(comparators.iter().map(std::convert::Into::into));
```

Comment:
Nit: I think that should work too and I find it more readable because it tells me what the expression is converting to.

```suggestion
                matches.extend(comparators.iter().map(HashableExpr::from));
```